src/rgw/rgw_rados.cc

```diff
@@ -7913,6 +7913,21 @@ int RGWRados::cls_obj_set_bucket_tag_timeout(RGWBucketInfo& bucket_info, uint64_
 }
 
 
+uint32_t RGWRados::calc_ordered_bucket_list_per_shard(uint32_t num_entries,
+						      uint32_t num_shards)
+{
+  constexpr uint32_t preferred_min_read = 10;
+  constexpr uint32_t read_factor = 4;
```

Comment:
See earlier comment about computing the expected factor

Reply:
if we were going to use that, it seems like we should pre-compute it

Reply:
@mattbenjamin We definitely could precompute the ln(shards)/shards portion whenever we reshard, and it's probably not a terrible idea on reshard to compute/store a factor for the 1000 entry page_size case since that's what will almost always be requested.

Having said that I'm not sure it will matter much.  We currently can return anywhere from 7000-25000 bucket entries (ie 7-25 pages) per second on a single NVMe backed OSD during ordered listing (tested up to 6.4M objects, likely slower with more).  I don't think computing this per page will add much to the latency until we speed up other aspects of bucket listing.

Reply:
that doesn't mean we should add cpu-intensive work to the call path

Reply:
if some additional cpu cycles in radosgw can save us a lot of read amp on the osds, i think it's worth doing. can we find a way to measure the read amp on some representative corpus of names, so we can compare this algorithm to some more naive ones?

Reply:
@cbodley sounds like we might not need a factor at all if we can just return a truncated set of results.  I'd vote going in that direction instead since it doesn't require any book keeping or complicated math.  

Also after speaking with Sage I don't think using the min is going to work well as it doesn't scale well to large numbers of shards (the expected minimum quickly hits 0 and then you have a divide by zero to deal with).  Something like max/average as the factor might work better.

I was curious how much overhead we already have so I ran a quick RGW ordered bucket list test with perf attached to the radosgw process.  This is listing 100,000 objects from a single bucket with 16 shards (ie 16x read-amp):

|  |  |
| -- | -- |
| Entries/Page | 1000 |
| Total Pages | 100 |
| Time | 6.1s |
| Pages/second | 16.4 |
| Total RadosGW Cycles During Test | 6,408,002,175 |
| RadosGW Cycles/Page | 64,080,022 |

The takeaway for me is that we shouldn't be afraid to spend a few extra cycles/page on some math if it saves us a lot of read-amp (and more than likely would actually reduce CPU usage by reducing read-amp!).

Reply:
> Also after speaking with Sage I don't think using the min is going to work well as it doesn't scale well to large numbers of shards (the expected minimum quickly hits 0 and then you have a divide by zero to deal with). Something like max/average as the factor might work better (if we can't truncate).

I'm not following this, @markhpc. If anything, having more shards than the requested entries (i.e., page size) argues for why the minimum is necessary.

Reply:
@ivancich I might not have been very clear, I mean the min formula I posted earlier to use as the factor:
```
factor = ((page_size / shards) / (page_size / shards) - sqrt(2 * page_size * ln(shards)/shards))
```

As the number of shards grows without an associated growth in page size you end up with the min at or below zero (which you normally would truncate to zero) but then you end up with a divide by zero you have to solve.

Instead I'm playing with using max/average which is probably not right but so far is working well enough and still grows as the number of shards grows, but not as aggressively.

```
  float read_factor = ((1.0 * num_entries / num_shards) + sqrt(2.0 * num_entries*log(num_shards)/num_shards)) / (1.0 * num_entries / num_shards);
```

Reply:
Ran some new tests with 64 buckets and 6.4M objects:

|  | Master | #30853 | #30853 + max_expr(1) |
| -- | -- | -- | -- |
| Entries/Page | 1000 | 1000 | 1000 |
| Total Pages | 6400 | 6400 | 6400 |
| Time | 778.0s | 288.7s | 272.9s |
| Pages/second | 8.23 | 22.17 | 23.45 |
| Total RadosGW Cycles During Test | 1,631,302,890,773 | 274,418,770,960 | 213,295,217,820 |
| RadosGW Cycles/Page | 254,891,077 | 42,877,933 | 33,327,379 |

Reply:
@markhpc sure, but log(shards) at the very least is for the given bucket a constant.  Throw me a bone here.